### PR TITLE
Add support for getting a recursive json file listing

### PR DESCRIPTION
**Pull request by ggoodman**
This PR adds a json listing of files based on content-negotiation.

The current behaviour remains untouched _except_ when a request for a directory has an `Accept` header that favours `json` over `html`. In this case, a new `TreeUtils` tool is used to construct and respond with a JSON representation of the contents of that version of the package. The recursion depth for this is configurable but defaults to `Number.MAX_VALUE`.

The format of the json looks like this:
```json
[
  {
    "path": "/.c9revisions",
    "lastModified": "2016-05-20T12:13:02.000Z",
    "mime": "application/octet-stream",
    "size": 136,
    "type": "directory",
    "children": [
      {
        "path": "/.c9revisions/README.md.c9save",
        "lastModified": "2013-03-26T21:59:05.000Z",
        "mime": "application/octet-stream",
        "size": 1233,
        "type": "file"
      },
      {
        "path": "/.c9revisions/index.js.c9save",
        "lastModified": "2013-03-26T21:59:22.000Z",
        "mime": "application/octet-stream",
        "size": 988,
        "type": "file"
      }
    ]
  },
  {
    "path": "/.npmignore",
    "lastModified": "2013-03-26T21:51:43.000Z",
    "mime": "text/plain",
    "size": 84,
    "type": "file"
  },
  {
    "path": "/README.md",
    "lastModified": "2013-03-26T21:59:04.000Z",
    "mime": "text/x-markdown",
    "size": 694,
    "type": "file"
  },
  {
    "path": "/index.js",
    "lastModified": "2013-03-26T21:59:22.000Z",
    "mime": "application/javascript",
    "size": 387,
    "type": "file"
  },
  {
    "path": "/package.json",
    "lastModified": "2013-03-26T22:00:38.000Z",
    "mime": "application/json",
    "size": 441,
    "type": "file"
  }
]
```

I considered making the depth configurable by query parameter but would like your feedback. This API would be used by Plunker as a replacement for its (deficient) package manager so that:

1. Users can find npm packages using the registry apis; then
2. Select which files from the selected package(s) they would like to import; then
3. Inject appropriate markup in `index.html` and `package.json` entries to refer to https://npmcdn.com.

### CODE REVIEW

**Comment by ggoodman** [comment]
@mjackson there is mutual recursion in `modules/TreeUtils.js` between `resolveDirectory` and `getEntries` that triggers the `no-use-before-define` eslint error.

To avoid this, I would need to deviate a bit from this project's style and use a non-`const` function assignment expression or use function statements. Thoughts?

**Comment by mjackson** [comment]
Thanks for the PR, @ggoodman :D This is great stuff.

My main concern with using content negotiation to serve this file is that we'll also need to use a `Vary` header to instruct the CDN to cache stuff based on the `Accept` header, which would mean a lot more cache misses.

Perhaps we could use a query parameter instead, something like `?json`. We could possibly enable this for individual files as well as directories. So you could get meta information about a single file if you needed it.

As for `no-use-before-define`, please feel free to just disable that in `.eslintrc`. I usually do.

**Comment by ggoodman** [comment]
@mjackson good point re-`Vary`. I've adjusted the PR with your comments and included a bonus bug-fix for determining if a directory redirect is needed in the presence of query parameters.

**Comment by ggoodman** [comment]
I also made a version of this suitable for running as a https://webtask.io in my [webtask brach](https://github.com/ggoodman/npm-http-server/tree/webtask) that _only_ does json directory listings. See: https://webtask.it.auth0.com/api/run/wt-ggoodman-gmail_com-0/npmcdn/genid/ or https://webtask.plunker.co/wt-ggoodman-gmail_com-0/npmcdn/genid/ (custom domain alias to the former).

File: modules/ResponseUtils.js [hunk 1]
```diff
@@ -10,6 +10,18 @@ export const sendText = (res, statusCode, text) => {
   res.end(text)
 }
 
+export const sendJSON = (res, json, maxAge = 0, statusCode = 200) => {
+  const text = JSON.stringify(json, null, 2)
```
Comment (mjackson) [inline modules/ResponseUtils.js]: Let's leave out whitespace here and just use `JSON.stringify(json)`. Will save us some bytes.

**Comment by mjackson** [comment]
This looks good, @ggoodman. Thanks for your work here. Just a few tweaks I'd like to make:

- Instead of returning a bare array in the response, let's wrap it in an object with a `files` property
- Instead of using `children` for the files in a directory, let's use `files` (mirrors the top-level prop)
- Instead of using `mime` let's use `contentType`

So, your example would look like this:

```json
{
  "files": [
    {
      "path": "/.c9revisions",
      "lastModified": "2016-05-20T12:13:02.000Z",
      "contentType": "application/octet-stream",
      "size": 136,
      "type": "directory",
      "files": [
        {
          "path": "/.c9revisions/README.md.c9save",
          "lastModified": "2013-03-26T21:59:05.000Z",
          "contentType": "application/octet-stream",
          "size": 1233,
          "type": "file"
        },
        {
          "path": "/.c9revisions/index.js.c9save",
          "lastModified": "2013-03-26T21:59:22.000Z",
          "contentType": "application/octet-stream",
          "size": 988,
          "type": "file"
        }
      ]
    },
    {
      "path": "/.npmignore",
      "lastModified": "2013-03-26T21:51:43.000Z",
      "contentType": "text/plain",
      "size": 84,
      "type": "file"
    },
    {
      "path": "/README.md",
      "lastModified": "2013-03-26T21:59:04.000Z",
      "contentType": "text/x-markdown",
      "size": 694,
      "type": "file"
    },
    {
      "path": "/index.js",
      "lastModified": "2013-03-26T21:59:22.000Z",
      "contentType": "application/javascript",
      "size": 387,
      "type": "file"
    },
    {
      "path": "/package.json",
      "lastModified": "2013-03-26T22:00:38.000Z",
      "contentType": "application/json",
      "size": 441,
      "type": "file"
    }
  ]
}
```

Also, let's leave the whitespace out of the JSON response.

**Comment by kentcdodds** [comment]
This is super duper cool. Looking forward to what else could be made possible because of this awesome addition to an already fantastic service.

File: modules/ResponseUtils.js (hunk 1 above)
Comment (ggoodman) [inline modules/ResponseUtils.js]: Adjusted.

**Comment by ggoodman** [comment]
@mjackson I've addressed your comments in the above commits.

**Comment by mjackson** [comment]
Thanks, @ggoodman! I'll roll this out today.

**Comment by ggoodman** [comment]
Thanks for being open to this and for the timely feedback :D

**Pull request merged**